mozjs-sys/build.rs

```diff
@@ -791,22 +816,53 @@ fn decompress_static_lib(archive: &Path, build_dir: &Path) -> Result<(), std::io
     let tar_gz = File::open(archive).unwrap_or({
         let mut workspace_dir = get_cargo_target_dir(build_dir).unwrap().to_path_buf();
         workspace_dir.pop();
-        File::open(workspace_dir.join(archive)).unwrap()
+        File::open(workspace_dir.join(archive))?
     });
     let tar = GzDecoder::new(tar_gz);
     let mut archive = Archive::new(tar);
     archive.unpack(build_dir)?;
     Ok(())
 }
 
-/// Download static library tarball instead of building it from source.
-fn download_static_lib_binaries(archive: &Path, build_dir: &Path) {
-    // Only download the files if build directory doesn't exist.
-    if !build_dir.exists() {
-        // TODO download from https
-        decompress_static_lib(archive, build_dir).expect("Failed to decompress static libs");
+/// Download spidermonkey archive by curl with provided base url. If it's None, it will use
+/// servo/mozjs's release page as base url.
+fn download_archive(base: Option<&str>) -> Result<PathBuf, std::io::Error> {
+    let base = base.unwrap_or("https://github.com/servo/mozjs/releases/download");
+    let version = env::var("CARGO_PKG_VERSION").unwrap();
+    let target = env::var("TARGET").unwrap();
+    let archive_path = PathBuf::from(env::var_os("OUT_DIR").unwrap()).join("libmozjs.tar.gz");
+    if !archive_path.exists() {
+        if !Command::new("curl")
+            .arg("-L")
+            .arg("-f")
+            .arg("-s")
+            .arg("-o")
+            .arg(&archive_path)
+            .arg(format!(
+                "{base}/mozjs-sys-v{version}/libmozjs-{target}.tar.gz"
+            ))
+            .status()?
+            .success()
+        {
+            return Err(std::io::Error::from(std::io::ErrorKind::NotFound));
+        }
     }
 
+    Ok(archive_path)
+}
+
+/// Link static library tarball instead of building it from source.
+fn link_static_lib_binaries(build_dir: &Path) -> Result<(), std::io::Error> {
+    if let Ok(archive) = env::var("MOZJS_ARCHIVE") {
```

Comment:
Could this be mapped to an `Option` instead of repeating the `download_archive` and `decompress_static_lib` calls below?

Reply:
How about something like this:

```rust
   let archive = let Ok(archive) = env::var("MOZJS_ARCHIVE") {
        download_archive(Some(&archive)).unwrap_or(PathBuf::from(archive));
    } else {
        download_archive(None)?;
    };
    decompress_static_lib(&archive, build_dir)?;
```

Reply:
I believe @sagudev wants `decompress_static_lib` in the first branch to have a hard fail because it manually specifies `MOZJS_ARCHIVE`, and we want it to panic instead of fall back.